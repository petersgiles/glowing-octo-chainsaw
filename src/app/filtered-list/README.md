# TEST

HELLO THIS IS README