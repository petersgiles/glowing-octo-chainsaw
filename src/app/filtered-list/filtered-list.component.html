<div class="sticky">
  <mdc-text-field [label]="label" [(ngModel)]="filter"></mdc-text-field>
</div>

<mdc-list twoLine>
  <mdc-list-item *ngFor="let item of filteredList" (click)="handleClick(item)" [class.selected]="item.id === selectedId" [routerLink]="item.link ? item.link : []">
    <mdc-list-item-text [secondaryText]="item.info ? item.info : ' '">{{item.title}}</mdc-list-item-text>
  </mdc-list-item>
</mdc-list>

