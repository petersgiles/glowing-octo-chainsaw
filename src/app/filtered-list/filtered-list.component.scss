.mda-list-item {
    cursor: pointer;
}

.mda-list-item:hover {
    background-color: #e3f2fd;
}

.mda-list-item-icon {
    position: absolute;
}

.mda-list-item-text {
    line-height: 1.2em;
    // margin-left: 70px;
}

.mda-list > .mda-list-item > .mda-list-item-text h4,
.mda-list > .mda-list-item > .mda-list-item-text h3 {
    line-height: 1.2em;
}

.sticky {
    position: sticky;
    top: 0;
    background-color: white;
}

.sticky h4 {
    padding: 24px 0;
    margin: 0;
}

.selected {
    background-color: #e3f2fd
}