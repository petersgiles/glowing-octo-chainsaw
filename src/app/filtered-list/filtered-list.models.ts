export interface ListItem<T> {
    id: string
    name: string
    entity?: T
    link?: string
    info?: string
}