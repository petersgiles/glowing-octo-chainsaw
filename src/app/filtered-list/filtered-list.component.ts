import { Component, Input, Output, EventEmitter } from '@angular/core'
import { ListItem } from './filtered-list.models'

@Component({
  selector: 'df-filtered-list',
  templateUrl: './filtered-list.component.html',
  styleUrls: ['./filtered-list.component.scss']
})

export class FilteredListComponent<T> {
  @Input() items: ListItem<T>[]
  @Input() label = "Filter"
  @Input() filterExpression = (listItem: ListItem<T>, filter: string) =>
    filter && listItem.name && listItem.name.toLowerCase().indexOf(filter.toLowerCase()) > -1
    || filter && listItem.info && listItem.info.toLowerCase().indexOf(filter.toLowerCase()) > -1

  @Output() select = new EventEmitter<ListItem<T>>()

  filter: string
  selectedId: string

  get filteredList() {
    if (!this.items) {
      return []
    }
    if (!this.filter) {
      return this.items
    }

    return this.items.filter(item =>
      this.filterExpression(item, this.filter)
    )
  }

  handleClick(item: ListItem<T>) {
    this.selectedId = item.id
    this.select.emit(item)
  }
}
