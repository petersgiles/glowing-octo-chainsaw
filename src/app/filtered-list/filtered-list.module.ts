import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { FilteredListComponent } from './filtered-list.component';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router'
import { MdcListModule, MdcTextFieldModule } from '@angular-mdc/web';

const COMPONENTS = [
    FilteredListComponent
]

@NgModule({
  imports: [
      CommonModule,
      FormsModule,      
      RouterModule.forChild([]),
      MdcListModule,
      MdcTextFieldModule
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class FilteredListModule { }