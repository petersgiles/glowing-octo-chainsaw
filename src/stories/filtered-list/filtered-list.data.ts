import { ListItem } from 'src/app/filtered-list';

export const SimpleFruits: ListItem<any>[] = [
    {
        id: "a",
        name: "Apple",
        info: "Malus pumila"
    },
    {
        id: "b",
        name: "Banana",
        info: "Musa acuminata"
    },
    {
        id: "c",
        name: "Cherry",
        info: "Prunus avium"
    }
]