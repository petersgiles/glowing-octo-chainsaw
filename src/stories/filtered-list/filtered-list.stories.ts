import { storiesOf, moduleMetadata } from '@storybook/angular';
import { withNotes } from '@storybook/addon-notes';
import { action } from '@storybook/addon-actions';
import { linkTo } from '@storybook/addon-links';

import { SimpleFruits } from './filtered-list.data'
import { FilteredListModule, ListItem } from 'src/app/filtered-list';
import { RouterModule } from '@angular/router';
import { APP_BASE_HREF } from '@angular/common';

storiesOf('Filtered List', module)
.addDecorator(
  moduleMetadata({
    imports: [
      FilteredListModule,
      RouterModule.forRoot([{
          path: 'iframe.html',
          redirectTo: ''
      }])
    ],
    providers: [
        {provide: APP_BASE_HREF, useValue: '/'}
    ]
  })
)
.add('Simple', () => {
  return ({
    template: `
    <df-filtered-list label="Fruits" [items]="fruits" (select)="handleSelectFruit($event)"></df-filtered-list>
  `,
    props: {
      fruits: SimpleFruits,
      handleSelectFruit: (item: ListItem<any>) => action("Fruit")(item.name)
    }
  })
});
