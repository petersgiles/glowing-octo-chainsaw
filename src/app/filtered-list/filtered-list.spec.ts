import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FilteredListComponent } from './filtered-list.component';
import { MdcTextFieldModule, MdcListModule, MdcListItem, MdcTextField } from '@angular-mdc/web';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { NO_ERRORS_SCHEMA, DebugElement } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { ListItem } from './filtered-list.models';
import { By } from '@angular/platform-browser';

const testItems: ListItem<any>[] = [
    {
        "id":  "6badd846-1232-4ccb-a031-93c72ce63795",
        "title":  "Apple",
        "info":  "Pyrus malus",
        "link": "/",
        "entity": { colour: "red" }
    },
    {
        "id":  "96f610a7-0999-4637-9001-774cc8b65cd2",
        "title":  "Apricot",
        "info":  "Prunus armeniaca"
    },
    {
        "id":  "4cae1e8d-ff43-400b-9170-6ab0c18691fd",
        "title":  "Avocado",
        "info":  "Persea americana"
    },
    {
        "id":  "c63dd826-b04b-40d3-9134-b4301d74a7af",
        "title":  "Banana",
        "info":  "Musa paradisicum"
    }
]

describe('FilteredListComponent', () => {
    let component: FilteredListComponent<any>;
    let fixture: ComponentFixture<FilteredListComponent<any>>;
    let firstItem: DebugElement

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            // schemas: [NO_ERRORS_SCHEMA],
            imports: [
                CommonModule,
                FormsModule,      
                RouterTestingModule,
                MdcListModule,
                MdcTextFieldModule
            ],
            declarations: [FilteredListComponent]
        })
            .compileComponents();
    }));

    beforeEach(() => {
        fixture = TestBed.createComponent(FilteredListComponent);
        component = fixture.componentInstance;
        component.items = testItems;
        fixture.detectChanges();

        firstItem = fixture.debugElement.query(By.directive(MdcListItem))
    });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should contain a list of items', () => {
        expect(fixture).toMatchSnapshot()
    });

    it('should emit an event and highlight when an item is selected', () => {
        spyOn(component.select, 'emit')
        firstItem.nativeElement.click()
        expect(component.select.emit).toHaveBeenCalledWith(testItems[0])
        
        fixture.detectChanges()
        expect(firstItem.nativeElement.classList.contains('selected')).toBe(true)
    });

    it('should contain a routerLink if link property is provided', () => {
        const routerLink = firstItem.nativeElement.getAttribute("ng-reflect-router-link")
        expect(routerLink).toBe(testItems[0].link)
    });

    it('should change the list when a filter is applied', () => {
        component.filter = 'Ap'
        fixture.detectChanges()
        expect(component.filteredList).toHaveLength(2)
    });

    it('should allow for custom entity filter expressions', () => {
        component.filterExpression = (listItem: ListItem<any>, filter: string) =>
            filter && listItem.entity && listItem.entity.colour.toLowerCase() === filter.toLowerCase()
        component.filter = 'red'
        fixture.detectChanges()
        expect(component.filteredList).toHaveLength(1)
    });
});
