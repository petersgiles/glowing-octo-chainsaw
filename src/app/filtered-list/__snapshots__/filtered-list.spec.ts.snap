// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredListComponent should contain a list of items 1`] = `
<df-filtered-list
  filterExpression={[Function Function]}
  items={[Function Array]}
  label={[Function String]}
  select={[Function EventEmitter]}
>
  <div
    class="sticky"
  >
    <mdc-text-field
      class="mdc-text-field ng-untouched ng-pristine ng-valid"
      ng-reflect-label="Filter"
    >
      
      <input
        aria-invalid="false"
        class="mdc-text-field__input"
        id="mdc-input-1"
        tabindex="0"
        type="text"
      />
      
      <label
        class="mdc-floating-label"
        for="mdc-input-1"
        mdcfloatinglabel=""
        ng-reflect-for="mdc-input-1"
      >
        Filter
      </label>
      
      <mdc-line-ripple
        class="mdc-line-ripple"
      />
      
    </mdc-text-field>
  </div><mdc-list
    aria-orientation="vertical"
    class="mdc-list mdc-list--two-line"
    ng-reflect-two-line=""
    role="list"
    twoline=""
  >
    
    <mdc-list-item
      class="mdc-list-item"
      id="mdc-list-item-4"
      ng-reflect-router-link="/"
      role="listitem"
      tabindex="-1"
    >
      <mdc-list-item-text
        class="mdc-list-item__text"
        ng-reflect-secondary-text="Pyrus malus"
      >
        
        <span
          class="mdc-list-item__primary-text"
        >
          Apple
        </span>
        
        <span
          class="mdc-list-item__secondary-text"
        >
          Pyrus malus
        </span>
      </mdc-list-item-text>
    </mdc-list-item>
    <mdc-list-item
      class="mdc-list-item"
      id="mdc-list-item-5"
      ng-reflect-router-link=""
      role="listitem"
      tabindex="-1"
    >
      <mdc-list-item-text
        class="mdc-list-item__text"
        ng-reflect-secondary-text="Prunus armeniaca"
      >
        
        <span
          class="mdc-list-item__primary-text"
        >
          Apricot
        </span>
        
        <span
          class="mdc-list-item__secondary-text"
        >
          Prunus armeniaca
        </span>
      </mdc-list-item-text>
    </mdc-list-item>
    <mdc-list-item
      class="mdc-list-item"
      id="mdc-list-item-6"
      ng-reflect-router-link=""
      role="listitem"
      tabindex="-1"
    >
      <mdc-list-item-text
        class="mdc-list-item__text"
        ng-reflect-secondary-text="Persea americana"
      >
        
        <span
          class="mdc-list-item__primary-text"
        >
          Avocado
        </span>
        
        <span
          class="mdc-list-item__secondary-text"
        >
          Persea americana
        </span>
      </mdc-list-item-text>
    </mdc-list-item>
    <mdc-list-item
      class="mdc-list-item"
      id="mdc-list-item-7"
      ng-reflect-router-link=""
      role="listitem"
      tabindex="-1"
    >
      <mdc-list-item-text
        class="mdc-list-item__text"
        ng-reflect-secondary-text="Musa paradisicum"
      >
        
        <span
          class="mdc-list-item__primary-text"
        >
          Banana
        </span>
        
        <span
          class="mdc-list-item__secondary-text"
        >
          Musa paradisicum
        </span>
      </mdc-list-item-text>
    </mdc-list-item>
  </mdc-list>
</df-filtered-list>
`;
